lib/internal/modules/esm/loader.js

```diff
@@ -194,36 +215,51 @@ class ESMLoader {
   /**
    * Collect custom/user-defined hook(s). After all hooks have been collected,
    * calls global preload hook(s).
-   * @param {object | object[]} customLoaders A list of exports from
-   * user-defined loaders (as returned by ESMLoader.import()).
+   * @param {Map<URL['href'], ExportedHooks} customLoaders
+   *  A list of exports from user-defined loaders (as returned by
+   *  ESMLoader.import()).
    */
   async addCustomLoaders(
-    customLoaders = [],
+    customLoaders = new SafeMap(),
   ) {
-    if (!ArrayIsArray(customLoaders)) customLoaders = [customLoaders];
-
-    for (let i = 0; i < customLoaders.length; i++) {
-      const exports = customLoaders[i];
+    // Maps are first-in-first-out, but hook chains are last-in-first-out,
+    // so create a new container for the incoming hooks (which have already
+    // been reversed).
+    const globalPreloaders = new SafeMap();
+    const resolvers = new SafeMap();
+    const loaders = new SafeMap();
+
+    for (const { 0: url, 1: exports } of customLoaders) {
       const {
         globalPreloader,
         resolver,
         loader,
       } = ESMLoader.pluckHooks(exports);
 
-      if (globalPreloader) ArrayPrototypePush(
-        this.#globalPreloaders,
+      if (globalPreloader) globalPreloaders.set(
+        url,
         FunctionPrototypeBind(globalPreloader, null), // [1]
       );
-      if (resolver) ArrayPrototypePush(
-        this.#resolvers,
+      if (resolver) resolvers.set(
+        url,
         FunctionPrototypeBind(resolver, null), // [1]
       );
-      if (loader) ArrayPrototypePush(
-        this.#loaders,
+      if (loader) loaders.set(
+        url,
         FunctionPrototypeBind(loader, null), // [1]
       );
     }
 
+    // Append the pre-existing hooks (the builtin/default ones)
+    for (const p of this.#globalPreloaders) globalPreloaders.set(p[0], p[1]);
+    for (const p of this.#resolvers) resolvers.set(p[0], p[1]);
+    for (const p of this.#loaders) loaders.set(p[0], p[1]);
+
+    // Replace the obsolete maps with the fully-loaded & properly sequenced one
+    this.#globalPreloaders = globalPreloaders;
+    this.#resolvers = resolvers;
+    this.#loaders = loaders;
```

Comment:
Could do, but that seems like reinventing the wheel. I don't need to arbitrarily manipulate items in the collection, just the overall sequence needs to be maintained.

I think Map is a more appropriate data structure because:

* non-sequential memory (not terribly important here, but true)
* items _can_ be accessed by name
* iteration is simpler

Lemme do this clean-up below and revisit?